### ISSUE TITLE
storage: figure out strategy for `KEY ... NOT ENFORCED` on Kafka sinks

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#19143]
### Feature request

Materialize today requires that it can verify that the `KEY` you specify in a Kafka sink is a true key, as assessed by the optimizer (e.g., due to the presence of a `GROUP BY` on the key).

In many cases, the optimizer is unable to prove this property, but the user "knows" it due to some property of their data. We currently support an undocumented `NOT ENFORCED` clause that allows the user to force Materialize to create a sink with an unverified key:

```sql
CREATE SINK ...
INTO KAFKA ...
KEY (col1, col2, ...) NOT ENFORCED
```

We repeatedly recommend this to users, and then wring our hands about the fact that this option is not documented. Some recent examples:

* https://materializeinc.slack.com/archives/CPNFV9CVB/p1683221718326749
* https://materializeinc.slack.com/archives/CU7ELJ6E9/p1664222165387019?thread_ts=1664220082.935699&amp;cid=CU7ELJ6E9

It's time to decide on a path forward. Are we comfortable enough with the `NOT ENFORCED` clause that we can document it? If not, what is the blessed workaround?

cc @aljoscha @morsapaes @jseldess @josharenberg 

### Work items

- [x] Add a test for #20008 using pgtest framework ([context](https://github.com/MaterializeInc/materialize/issues/19143#issuecomment-1599598792))
- [x] #13119
- [ ] Update the notice from [#20008](https://github.com/MaterializeInc/materialize/issues/20008) to include a link to the documentation

### Descoped work items

- ~Update syntax to match `SUBSCRIBE` ([context](https://github.com/MaterializeInc/materialize/issues/19143#issuecomment-1579856210))~ Moved to #20135.
- ~Get consensus on closing out https://github.com/MaterializeInc/materialize/issues/17549 ([context](https://github.com/MaterializeInc/materialize/issues/19143#issuecomment-1599593053))~ Leaving as future work.

**Comment by morsapaes** [comment]
Linking to #8200 so it doesn't get lost. That issue leads to a scenario where this workaround is required ([example](https://materializecommunity.slack.com/archives/C015KDVS7EV/p1685386584703889?thread_ts=1685371682.001959&cid=C015KDVS7EV)).

**Comment by sjwiesman** [comment]
I've come to the opinion that we bless `KEY() NOT ENFORCED` and document it as an official syntax. Unenforced keys are a well-understood concept by users, and every user I've told about this has immediately understood what it means and the implications if they get it wrong. 

What I don't want to see us do is go down a route where users are forced to instantiate an arrangement just to prove to use that a collection of columns are unique.

**Comment by benesch** [comment]
I'm comfortable exposing a knob like this but I do feel like the current syntax is in the wrong spot. I think it should match the new `SUBSCRIBE ... ENVELOPE UPSERT` syntax:

```
CREATE SINK ...
ENVELOPE UPSERT (
    KEY (..., ...) NOT ENFORCED
)
```

We don't make you type `NOT ENFORCED` for `SUBSCRIBE`, IIRC.

cc @aljoscha @andrewrodriguez-m

**Comment by benesch** [comment]
@dseisun-materialize was asking "why do we even have this restriction in the first place?" so I dredged up the historical context:

  * https://github.com/MaterializeInc/materialize/pull/6709/commits/564060eec609bf3c0f1e8c5f66a1ff61e70a8d7b
  * https://github.com/MaterializeInc/materialize/pull/6709

Figured it might be of interest more broadly.

**Comment by benesch** [comment]
> My personal opinion is that our current behaviour is the best among bad options, and we should go ahead and commit to and document it.

+1 on this, but I'm increasingly convinced we should take the opportunity to standardize the syntax with `SUBSCRIBE` first (https://github.com/MaterializeInc/materialize/issues/19143#issuecomment-1579856210).

**Comment by aljoscha** [comment]
> My personal opinion is that our current behaviour is the best among bad options, and we should go ahead and commit to and document it. (Though we should send a notice instead of just logging the warning so the user sees it.) But since I am the one who implemented the current behaviour I am a little biased here!

Yes! For both, the send-the-notice part is important before we commit to it.

**Comment by benesch** [comment]
As part of this work it's probably worth getting a test for #20008, too. It's relatively straightforward to do with the "pgtest" framework.

**Comment by benesch** [comment]
I added a list of the outstanding work items to the issue description.

**Comment by benesch** [comment]
> +1 on this, but I'm increasingly convinced we should take the opportunity to standardize the syntax with `SUBSCRIBE` first ([#19143 (comment)](https://github.com/MaterializeInc/materialize/issues/19143#issuecomment-1579856210)).

I had to change my tune about this. The `KEY` clause I've been griping about being in the wrong spot is already documented:

![image](https://github.com/MaterializeInc/materialize/assets/882976/08819df6-9605-40ad-a9d4-9f88141e695a)

albeit without `NOT ENFORCED`. Adding `NOT ENFORCED` to the documentation seems fine. It hardly makes the situation worse. I had thought the whole `KEY` clause was undocumented.

I put up a PR (https://github.com/MaterializeInc/materialize/pull/20196) that adds documentation and links to that documentation in the NOTICE response. That covers all the remaining work items in this epic. (I did edit the epic work items to explicitly punt on two work items. See the issue description for justification.)
